File: app/src/main/java/com/abhishek/germanPocketDictionary/utilities/SharedPreferenceManager.java
package com.abhishek.germanPocketDictionary.utilities;

import android.content.Context;
import android.content.SharedPreferences;

import androidx.preference.PreferenceManager;

import com.abhishek.germanPocketDictionary.model.Word;
import com.google.gson.Gson;
import com.google.gson.reflect.TypeToken;

import java.lang.reflect.Type;
import java.util.List;

import static com.abhishek.germanPocketDictionary.utilities.Utils.getJsonFromList;

/**
 * @author Abhishek Saxena
 * @since 19/4/20 11:19 AM
 */

public class SharedPreferenceManager {

    private static SharedPreferenceManager sInstance;
    //private Context context;
    private static SharedPreferences sharedPreferences;

    public static SharedPreferenceManager getInstance(Context context) {
        if (sInstance == null)
            sInstance = new SharedPreferenceManager(context);

        return sInstance;
    }

    private SharedPreferenceManager(Context context) {
        sharedPreferences = PreferenceManager.getDefaultSharedPreferences(context);
    }

    public <T> void setList(String key, List<T> list) {
        String json = getJsonFromList(list);
        if (json != null && !json.equals("")) {
            SharedPreferences.Editor editor = sharedPreferences.edit();
            editor.putString(key, json);
            editor.apply();
        }
    }

    public boolean contains(String key) {
        return sharedPreferences.contains(key);
    }

    public List<Word> getListFromPreference(String key) {
        Gson gson = new Gson();
        String json = sharedPreferences.getString(key, "");
        Type type = new TypeToken<List<Word>>() {
        }.getType();
        return gson.fromJson(json, type);

    }

    public void registerOnSharedPreferenceChangeListener(SharedPreferences.OnSharedPreferenceChangeListener listener){
        sharedPreferences.registerOnSharedPreferenceChangeListener(listener);
    }

    public void unregisterOnSharedPreferenceChangeListener(SharedPreferences.OnSharedPreferenceChangeListener listener){
        sharedPreferences.unregisterOnSharedPreferenceChangeListener(listener);
    }
}

File: app/src/main/java/com/abhishek/germanPocketDictionary/core/utils/SharedPreferenceManager.kt
package com.abhishek.germanPocketDictionary.core.utils

import android.content.Context
import android.content.SharedPreferences
import androidx.preference.PreferenceManager
import dagger.hilt.android.qualifiers.ApplicationContext
import javax.inject.Inject

class SharedPreferenceManager @Inject constructor(
    @ApplicationContext
    context: Context,
) {

    private val sharedPreferences: SharedPreferences? =
        PreferenceManager.getDefaultSharedPreferences(context)

    private val editor: SharedPreferences.Editor?
        get() = sharedPreferences?.edit()

    fun putBoolean(key: String, value: Boolean) {
        editor?.putBoolean(key, value)?.apply()
    }

    fun getBoolean(key: String): Boolean {
        return sharedPreferences?.getBoolean(key, false) ?: false
    }
}
